### ISSUE TITLE
Copy Event does not create Unique Identifier

### ISSUE THREAD

**Issue by mcsSolutions** [HomeSeer/Plugin-SDK#214]
Applicable to all HS4 and HS3 versions through 4.2

The copy event user action creates a new event with a name being appended with "Copy".  It however, does not create a UID to distinguish it from the event that was copied.

The resulting effect is that when a plugin evaluates the event trigger the same UID exists for multiple events so the plugin will fire multiple event triggers.  An extract below shows where a plugin gets the list of triggers associated with the plugin and then iterates through the list.  iUID is the one that satisfied the plugin's trigger criteria.  Since it exists in the cloned event and the original event it will fire the trigger for both of these two events, while only one of the two actually satisfied the trigger criteria.

[code]
                        Dim configuredTriggers As HomeSeer.PluginSdk.Events.TrigActInfo() = hs.GetTriggersByType(PLUGIN_NAME, 1)
                        For Each configuredTrigger As HomeSeer.PluginSdk.Events.TrigActInfo In configuredTriggers
                            'hsWritelog(PLUGIN_DEBUG, "Trigger Fire UID " & iUID.ToString & ", TrigListUID=" & configuredTrigger.UID.ToString & configuredTrigger.)
                            If iUID = configuredTrigger.UID Then
                                hs.TriggerFire(PLUGIN_NAME, configuredTrigger)
                                hsWritelog(PLUGIN_DEBUG, "Trigger Fired for Event Ref " & configuredTrigger.evRef.ToString & " UID " & iUID.ToString)
                                'DebugPrint("Fire " & iUID.ToString & " " & configuredTrigger.evRef.ToString & " " & configuredTrigger.ToString)
                            End If
                        Next
[/code]

**Comment by mcsSolutions** [comment]
The lack of a unique ID across events is likely not a real issues if the developer develops their own unique ID as some composite of the event reference and the ID.

**Comment by spudwebb** [comment]
We already discussed this topic some time ago in the forums: https://forums.homeseer.com/forum/lighting-primary-technology-plug-ins/lighting-primary-technology-discussion/mcsmqtt-michael-mcsharry/1257398-copying-events-using-mcsmqtt-is-broken?p=1257632#post1257632
For HS3 plugins, there is some workarounds. Nonetheless I had a look at HS source code to see what it does to copy an event. It turns out that for actions a new unique ID is generated, but not for triggers. there is a comment saying that generating a new unique ID for triggers would break some plugins (including the ZWave plugin)

I also tested copy event with HS4 plugins, and according to my tests it doesn't work at all, neither for triggers nor actions. The HS4 plugin actions and triggers built using the SDK with JUI views are not copied because internally the PluginSdk build a page with an ID being  $"{_eventRef}-{_id}", then when the trigger or action is copied the new event ref does no longer match what was stored in the page ID, so an exception is raised in the setter of the ConfigPage property.
I created issue PSDK-212 for this problem.

```
        protected Page ConfigPage {
            get => _configPage ?? (_configPage = PageFactory.CreateEventActionPage(PageId, Name).Page);
            set {
                if (value.Id != PageId) {
                    throw new Exception($"The page ID must be {PageId}");
                }

                if (value.Type != EPageType.EventAction) {
                    throw new ViewTypeMismatchException("The page type must be EPageType.EventAction");
                }

                _configPage = value;
            }
        }
```

**Comment by spudwebb** [comment]
This issue is fixed in HS 4.2.0.5, new UID are not created for triggers when the "Copy Event" function is used.
And for HS4 plugins all occurrences of the PageId  string ($"{_eventRef}-{_id}")  are replaced with the new event ref and action/trigger UID in the new event.
